### PR TITLE
feat: Automatically populate uuid4 fields

### PR DESCRIPTION
**Pull request by parthea**
Adds support for automatically populating uuid fields as per https://google.aip.dev/client-libraries/4235

Towards b/322910372

### CODE REVIEW

**Review by vchudnov-g — COMMENTED** [review]
Looks good. Only minor comments in the files, but one overall comment before I approve: could we have some sort of golden file that exercises the `uuid` tests you added? (i.e. from a proto with an autopopulated field)

File: gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2
Comment (parthea) [reply gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2]: a) Good catch! I've fixed this in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/6bff6d8bda7b7f84dc269e44e5d496c468a0694e and https://github.com/googleapis/gapic-generator-python/pull/1985/commits/108ad38668d1b736cb0b8a44c93af0d226ca64bd by setting the `default` argument of [jinja-filters.map](https://jinja.palletsprojects.com/en/3.0.x/templates/#jinja-filters.map). I've also opened https://github.com/googleapis/gapic-generator-python/pull/1989 to fix other usages in the code.
b) I've confirmed that the list is empty if none of the attributes are set and confirmed via the jinja docs that [jinja-filters.map](https://jinja.palletsprojects.com/en/3.0.x/templates/#jinja-filters.map) can be used for looking up attributes.

File: gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2
Comment (parthea) [reply gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/8bb4496ae83ca66776b841646a6b9e7b0afed6d0

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2
Comment (parthea) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/commit/8bb4496ae83ca66776b841646a6b9e7b0afed6d0

File: gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2
Comment (parthea) [reply gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:528]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/1c042c015847f1b2baf71c2ff5762851e586d1d1

File: gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2
Comment (parthea) [reply gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/11bb6b4080e7900c24c7269fb910eb598236c398

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (parthea) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/11bb6b4080e7900c24c7269fb910eb598236c398

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2
Comment (parthea) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2:271]: The ads team is moving away from custom templates. Having common macros between the ads and standard templates would result in additional clean up once we remove the ads templates.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2
Comment (parthea) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2:279]: I agree, I applied your suggestion.

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (parthea) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/bbb95f085cf3b32ed33b8a320b1f7a68b50a1509

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2
Comment (parthea) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2:280]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/7fa920b03341ed329642546e47a49415d4d1081c

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (parthea) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/7d9b35c998edde74c4638d17572dd1448e87141e

**Comment by parthea** [comment]
> could we have some sort of golden file that exercises the uuid tests you added? (i.e. from a proto with an autopopulated field)

This is already being done since we're running tests against `gapic-showcase` version 0.32.0. See the `showcase-unit` presubmits.
https://github.com/googleapis/gapic-generator-python/blob/24a23a1ab885246e447c0010b2e5602209dfbb8d/noxfile.py#L197

On an unrelated note, I've opened PR https://github.com/googleapis/gapic-generator-python/pull/1991 to add showcase to the `goldens` directory but this does not add value in terms of unit testing as this is already being done.

File: gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2
Comment (vchudnov-g) [inline gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:661]: We have this check in multiple places, so I suggest having this in a macro. Even if we have to have duplicate macros for the ads vs non-ads templates (we can define the Ads macro in this file, IIUC)

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (vchudnov-g) [inline gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2:147]: Ditto previous comment: We have this check in multiple places, so I suggest having this in a macro.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2:271]: But until they do move away, it's duplicate code. If we can have ads import a macro file that lives in the standard template directory (or do we need a symlink?) then when ads removes their templates, there will be nothing to do. At worst we wind up with two macro files: one that was intended only for standard templates, one that was intended for both templates.

I think this would be more elegant, particularly if the Ads migration takes a while. But I won't belabor the point or block on this; it could hurt us only of the Ads migration takes a while and we need to maintain both locations.

File: gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2
Comment (vchudnov-g) [reply gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2]: Glad about (a)

Thanks for confirming (b). It wasn't clear to me from the Jinja docs (and it still isn't) that if the `map` filter returns `None` (or `""`, the default you've set) the value wouldn't be included in the list (as opposed to getting a list of `None`s or of `""`s)

**Review by vchudnov-g — APPROVED** [review]
Thanks for doing this!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2
Comment (parthea) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/_client_macros.j2:271]: I've created https://github.com/googleapis/gapic-generator-python/issues/1994 to follow up on whether we want to try to remove the ads templates (working with the Ads team), or refactor code so it can be shared between both template directories. Let's discuss with the ads team to understand the gaps in the standard templates before spending time on refactoring code. I've filed https://github.com/googleapis/gapic-generator-python/issues/1994 to follow up on this.

File: gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2
Comment (parthea) [reply gapic/ads-templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/70f0c64b54bb0551295dbf7f23ced7507202044c

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2
Comment (parthea) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_macros.j2]: Fixed in https://github.com/googleapis/gapic-generator-python/pull/1985/commits/70f0c64b54bb0551295dbf7f23ced7507202044c

**Pull request merged**
